### PR TITLE
[SM-64] Soft Delete Secrets

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to provide an endpoint in the API to bulk soft delete secrets for the Secrets Manager project.

## Code changes

- bitwarden_license/src/Commercial.Core/SecretManagerFeatures/SecretManagerCollectionExtensions.cs
bitwarden_license/src/Commercial.Core/SecretManagerFeatures/Secrets/DeleteSecretCommand.cs
src/Core/SecretManagerFeatures/Secrets/Interfaces/IDeleteSecretCommand.cs

Command for bulk soft deleting secrets

- bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/SecretRepository.cs
src/Core/Repositories/ISecretRepository.cs

Adding in a GetManyByIds method for the Secret repo.

- bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/Secrets/DeleteSecretCommandTests.cs
test/Api.Test/Controllers/SecretsControllerTests.cs

Unit tests for new feature.

- src/Api/Controllers/SecretsController.cs

New endpoint on the controller.

- src/Api/SecretManagerFeatures/Models/Response/SecretDeleteBulkResponseModel.cs

Response model to be return to client.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by Hinton — COMMENTED** [review]
Can we add an integration test for this endpoint?

File: src/Api/Controllers/SecretsController.cs [hunk 1]
```diff
@@ -60,5 +62,14 @@ public async Task<SecretResponseModel> UpdateSecretAsync([FromRoute] Guid id, [F
             var result = await _updateSecretCommand.UpdateAsync(updateRequest.ToSecret(id));
             return new SecretResponseModel(result);
         }
+
+        // TODO Once permissions are setup for Secrets Manager need to enforce them on delete.
+        [HttpDelete("secrets/bulk")]
```
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Api/Controllers/SecretsController.cs:67]: in principle, probably, but the reason is really if some load balancer or client we care about doesn't support it.

I see some chatter than openapi dropped DELETE bodies [for a while](https://stackoverflow.com/a/53228595), but I see them present in our sdk api builds so 🤷 

In general we define a post alongside deletes, so I'm OK with moving that convention. DELETE has always been a bit of a pain.

File: bitwarden_license/src/Commercial.Core/SecretManagerFeatures/Secrets/DeleteSecretCommand.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            var secrets = await _secretRepository.GetManyByIds(ids);
+
+            if (secrets?.Any() != true)
+            {
+                throw new NotFoundException();
+            }
+
+            var results = new List<Tuple<Guid, string>>();
+
+            foreach (Guid id in ids)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretManagerFeatures/Secrets/DeleteSecretCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2253/commits/eb491fdf7f881a8af3b48428b6f7218f9db6e148

File: src/Api/Controllers/SecretsController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply src/Api/Controllers/SecretsController.cs]: `[HttpPost("secrets/bulk")]`

Reads like bulk creating secrets. 

went with `[HttpPost("secrets/delete")]` instead

File: bitwarden_license/src/Commercial.Core/SecretManagerFeatures/SecretManagerCollectionExtensions.cs [hunk 3]
```diff
@@ -13,6 +13,8 @@ public static void AddSecretManagerServices(this IServiceCollection services)
             services.AddScoped<ICreateSecretCommand, CreateSecretCommand>();
             services.AddScoped<IUpdateSecretCommand, UpdateSecretCommand>();
             services.AddScoped<IDeleteSecretCommand, DeleteSecretCommand>();
+            services.AddScoped<ICreateProjectCommand, CreateProjectCommand>();
+            services.AddScoped<IUpdateProjectCommand, UpdateProjectCommand>();
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretManagerFeatures/SecretManagerCollectionExtensions.cs:17]: These shouldn't be added right?

**Review by Hinton — APPROVED** [review]

**Pull request merged**